Prevent device (de)assignment to dom0

XenServer is built to require deassignment of devices to the quarantine before
being assigned elsewhere. However, there is nothing stopping a malicious
userspace from deassigning a device with ongoing DMA operations to the
quarantine domain and then immediately back to dom0 to potentially corrupt or
attack the dom0 kernel.

So, when Xen is locked down, prevent device assignment and deassignment to dom0.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
Signed-off-by: Frediano Ziglio <frediano.ziglio@citrix.com>

diff --git a/xen/drivers/passthrough/pci.c b/xen/drivers/passthrough/pci.c
index 9c16dfdf22f3..bb935ac126a5 100644
--- a/xen/drivers/passthrough/pci.c
+++ b/xen/drivers/passthrough/pci.c
@@ -35,6 +35,7 @@
 #include <xen/vpci.h>
 #include <xen/msi.h>
 #include <xen/lockdown.h>
+#include <xen/efi.h>
 #include <xsm/xsm.h>
 #include "ats.h"
 
@@ -973,8 +974,10 @@ static int deassign_device(struct domain *d, uint16_t seg, uint8_t bus,
 
         target = dom_io;
     }
-    else
+    else if ( !efi_secure_boot )
         target = hardware_domain;
+    else
+        return -EPERM;
 
     while ( pdev->phantom_stride )
     {
@@ -1645,6 +1648,9 @@ static int assign_device(struct domain *d, u16 seg, u8 bus, u8 devfn, u32 flag)
     if ( !arch_iommu_use_permitted(d) )
         return -EXDEV;
 
+    if ( efi_secure_boot && d == hardware_domain )
+        return -EPERM;
+
     /* device_assigned() should already have cleared the device for assignment */
     ASSERT(pcidevs_locked());
     pdev = pci_get_pdev(NULL, PCI_SBDF(seg, bus, devfn));
